iaia/magicmodule.py

```diff
@@ -56,27 +59,45 @@ def __repr__(self):
     def __call__(self, *args, **kw):
         key = self.call_key(*args, **kw)
         if key not in self.funcs:
-            self.make_function(*args, **kw)
+            self.make_functions(*args, **kw)
         exc = None
-        try:
-            return self.funcs[key](*args, **kw)
-        except Exception as e:
-            exc = e
-        print(f"Attempting to fix exception {exc}...")
-        self.fix_function(exc, *args, **kw)
-        return self.funcs[key](*args, **kw)
+        for func_i, func in self.funcs[key].items():
+            try:
+                return func(*args, **kw)
+            except Exception as e:
+                exc = e
+            print(f"Attempting to fix exception {exc} in func {func_i}...")
+            func = self.fix_function(func_i, exc, *args, **kw)
+            return func(*args, **kw)
 
     def call_key(self, *args, **kw):
         return tuple([len(args), *sorted(kw.keys())])
 
+    def format_type_for_prompt(self, arg):
+        """
+        Format arguments for the prompt, adding info about function args
+        """
+        name = type(arg).__name__
+        if name == "function":
+            arg_count = arg.__code__.co_argcount
+            return "Callable[" + ", ".join(["Any"] * arg_count) + "]"
```

Comment:
As a followup this could probably use `inspect.signature()` on the method to get both annotations/types and argument names. Though I guess there's not room enough for anything but positional argument types in Callable. Maybe another option is to make it a multi-line signature and include a comment after Callable, like:

```python
def func(
  x: Callable[[int], int], # callable with signature {inspect.signature(func)}
  ...
):
```

Reply:
Another option that could be useful is to start the signature with a docstring, like:

```python
def func(x: Callable[int], ...):
    """
    x: a function of the form {inspect.signature(func)}
```

(Not closing the docstring so that GPT can still add to it)

Reply:
Went with the second option after some deliberation, but a bit worried that this edge case of callable functions will confuse the AI since docstrings don't normally start with argument explanations. But a docstring potentially allows for improvements down the road using comments surrounding the call site.

Other way seems less pythonic, too.